### PR TITLE
Fix aliasing annotation for aten.dropout

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116998

It can return an alias to its input:
```python
>>> import torch
>>> x = torch.randn(100)
>>> torch.ops.aten.dropout(x, train=False, p=0.0) is x
True
>>> 
```

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -296,7 +296,7 @@
 
 - func: _shape_as_tensor(Tensor self) -> Tensor
 
-- func: dropout(Tensor input, float p, bool train) -> Tensor
+- func: dropout(Tensor(a) input, float p, bool train) -> Tensor(a)
```
Comment (jerryzh168) [inline aten/src/ATen/native/native_functions.yaml:299]: does this mean the output tensor is optionally aliased with `input`?

Comment (jansel) [reply aten/src/ATen/native/native_functions.yaml:299]: I think so

Comment (tugsbayasgalan) [reply aten/src/ATen/native/native_functions.yaml:299]: I feel like this won't work for QAT because there will be no way to pattern match anymore....

**Comment by vadimkantorov** [comment]
Related discussion on making explicit sematnics for "possibly alias" / "possibly new tensor": 
- https://github.com/pytorch/pytorch/issues/106584#issuecomment-1666483832

I think dropout is also one of the ops which can either return a new tensor or return input direcly (or a clone), depending on selected semantics - and all of them might make sense. So it's important to document clearly the existing alias semantics, and if need to support/express more modes, might be a good idea to introduce an explicit argument for this...

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/native/native_functions.yaml:299]: yeah wondering if we can decompose this to a different op/ops for QAT, although the decomposition might change the semantics
